# Did Chester rig the election?
### lets find out

In [1]:
import pandas as pd
import json

## We use twitch chat data for subber sunday stream from 4:33:27 to 4:37:30
## data is a json format

In [11]:
open1= open('D:/downloads/release/chat data/slacks_2.json', 'r', encoding='utf-8')

In [12]:
json1 = json.load(open1)

In [86]:
list_time = []
list_name = []
list_chat = []
list_sub = []
list_test = []

for i in range(0, len(json1['comments'])):
    list_time.append(json1['comments'][i]['created_at'])
    list_name.append(json1['comments'][i]['commenter']['display_name'])
    list_chat.append(json1['comments'][i]['message']['body'])
    int_sub = 0
    if 'user_badges' in json1['comments'][i]['message']:
        for j in range(0,len(json1['comments'][i]['message']['user_badges'])):
            if 'subscriber' in json1['comments'][i]['message']['user_badges'][j].values():
                int_sub = 1
    list_sub.append(int_sub)

In [87]:
df = pd.DataFrame({'time':list_time, 'name':list_name,'chat':list_chat,'sub':list_sub})

## Lets see what we are working with 

In [88]:
df.head()

time            name  \
0  2020-11-15T22:07:19.887Z     spacewolf73   
1  2020-11-15T22:07:20.094Z  StreamElements   
2  2020-11-15T22:07:22.296Z            Dopp   
3   2020-11-15T22:07:22.79Z    ChesterLibra   
4  2020-11-15T22:07:22.886Z         Jolixia   

                                                chat  sub  
0                                             !prime    1  
1  Are you a slacksFAN ? Enjoying the stream? Hav...    1  
2                                            !vote 2    1  
3                            @sheepsticked PepeHands    1  
4                                            !vote 2    1

### column 'sub': 1 = sub, 0 = not a sub

### Lets find out if someone voted more than once and tried to commit voter fraud:

In [115]:
pd.DataFrame(df[df['chat'].str.contains('!vote ')]['name'].value_counts())['name'].value_counts()

1    50
2    11
3     1
Name: name, dtype: int64

In [144]:
list_fruad = list(pd.DataFrame(df[df['chat'].str.contains('!vote ')]['name'].value_counts())['name'].value_counts())

In [147]:
print('Number of people that voted more than once: ' + str(list_fruad[1]+list_fruad[2]))

Number of people that voted more than once: 12



### there is a clear sign of voter fraud, 12 people tried to vote more than once!


### Also some people voted for a 3rd option, lets remove them shall we

In [117]:
df_vote = df[df['chat'].str.contains('!vote ')]

In [127]:
df_vote = df_vote[df_vote['chat']!='!vote 3'].reset_index(drop = True)

In [130]:
df_vote2 = df_vote.groupby('name', as_index=False).first()


### Lets take a look at how people voted after removing their additional votes with the table below


In [142]:
df_vote2

name                      time     chat  sub
0         Angel2551  2020-11-15T22:09:42.745Z  !vote 1    0
1          Attyla11  2020-11-15T22:08:23.803Z  !vote 2    1
2    BaconHazardous  2020-11-15T22:07:38.117Z  !vote 1    1
3      Barlakopofai   2020-11-15T22:07:48.25Z  !vote 1    0
4      ChesterLibra  2020-11-15T22:07:27.188Z  !vote 2    1
..              ...                       ...      ...  ...
56  thehooliganshow  2020-11-15T22:07:34.565Z  !vote 2    0
57            vailo  2020-11-15T22:07:41.932Z  !vote 2    1
58      wheatiesbox  2020-11-15T22:07:32.519Z  !vote 1    1
59      whiskey_dog  2020-11-15T22:07:40.222Z  !vote 2    1
60      zalathormnw  2020-11-15T22:07:27.778Z  !vote 2    1

[61 rows x 4 columns]

In [131]:
df_vote2['chat'].value_counts()

!vote 2    39
!vote 1    22
Name: chat, dtype: int64


### It looks like 39 people voted 2 and 22 people voted 1

### Now lets increase the weightage of votes from subs and see what we get


In [137]:
int_cool_meepo = 0
int_lore = 0

for i in range(0,len(df_vote2)):
    if df_vote2['chat'][i] == '!vote 1':
        if df_vote2['sub'][i] == 1:
            int_cool_meepo +=2
        else:
            int_cool_meepo +=1
    if df_vote2['chat'][i] == '!vote 2':
        if df_vote2['sub'][i] == 1:
            int_lore +=2
        else:
            int_lore +=1

In [148]:
print('votes for Cooler Meepo: ' + str(int_cool_meepo))

votes for Cooler Meepo: 35


In [150]:
print('votes for Lore: ' + str(int_lore))

votes for Lore: 60


In [154]:
print('Percentage of votes for the lore emote: '+str((int_lore/(int_lore+int_cool_meepo))*100)+'%')

Percentage of votes for the lore emote: 63.1578947368421%


In [155]:
print('Percentage of votes for the cooler meepo emote: '+str((int_cool_meepo/(int_lore+int_cool_meepo))*100)+'%')

Percentage of votes for the cooler meepo emote: 36.84210526315789%


### The lore emote won with approximately 63% votes
### Chester did not rig the election